src/jit/lower.cpp

```diff
@@ -338,24 +338,96 @@ void Lowering::DecomposeNode(GenTreePtr* pTree, Compiler::fgWalkData* data)
     case GT_CAST:
         {
             assert(tree->gtPrev == tree->gtGetOp1());
-            NYI_IF(tree->gtOverflow(), "TYP_LONG cast with overflow");
-            switch (tree->AsCast()->CastFromType())
+            GenTree* srcOp = tree->gtGetOp1();
+            var_types srcType = tree->AsCast()->CastFromType();
+            var_types dstType = tree->AsCast()->CastToType();
+
+            if (tree->gtFlags & GTF_UNSIGNED)
             {
-            case TYP_INT:
-                if (tree->gtFlags & GTF_UNSIGNED)
+                srcType = genUnsignedType(srcType);
+            }
+
+            if (genActualTypeIsIntOrI(srcType) || varTypeIsGC(srcType))
```

Comment:
OK, I now see that this was actually intended to exclude long types, so I take it back. A comment might be good to clarify that point (i.e. that we're looking for source types that fit in a single 32-bit register).